helix-core/src/main/java/org/apache/helix/controller/stages/MessageGenerationPhase.java

```diff
@@ -271,6 +271,23 @@ private void generateMessage(final Resource resource, final BaseControllerDataPr
     } // end of for-each-partition
   }
 
+  private boolean shouldCreateSTCancellation(Message pendingMessage, String currentState,
+      String desiredState, String initialState) {
+    if (pendingMessage == null) {
+      return false;
+    }
+    if (NO_DESIRED_STATE.equals(desiredState)) {
+      return true;
+    }
+
+    // Don't cancel the ST for below scenarios:
+    // 1. current state has arrived at pending message's to state
+    // 2. pending message is an ERROR reset: ERROR -> initState (eg. OFFLINE)
+    return !currentState.equalsIgnoreCase(pendingMessage.getToState())
+        && !HelixDefinedState.ERROR.toString().equals(pendingMessage.getFromState())
+        && !initialState.equals(pendingMessage.getToState());
```

Comment:
Question: This line logic is expected. But I did not find it from previous code?

Reply:
You meant this? https://github.com/apache/helix/blob/b1b194c1bffa68a53f1a81ba307452474c71a016/helix-core/src/main/java/org/apache/helix/controller/stages/MessageGenerationPhase.java#L214-L216

Previously only `!currentState.equalsIgnoreCase(pendingMessage.getToState()` is checked. The ERROR state and initial state are not.

Reply:
Minor thing, as we discussed offline, the comment is about why we don't cancel, but the code returns true if the message should be canceled. Please try to make them align, or this part tends to be confusing.

Reply:
There is a typo in the code,
(HelixDefinedState.ERROR.name().equals(pendingMessage.getFromState()) && initialState.equals(pendingMessage.getToState()) should be checked together.

Reply:
@jiajunwang Oh...Good catch! Yes. `(HelixDefinedState.ERROR.name().equals(pendingMessage.getFromState()) && initialState.equals(pendingMessage.getToState())` is supposed together as it represents `ERROR -> OFFLINE`.